<a href="https://colab.research.google.com/github/Thiva02/Statistical-Learning-e22399/blob/main/Statistical_Learning_Assignement_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Q3)

### **1.**
A probability space consists of three components:

* **Sample Space ($\Omega$):** Since the random variable $X$ takes values in $\{0, 1, 2, 3\}$, all possible outcomes,

  $$\Omega = \{ \omega_0, \omega_1, \omega_2, \omega_3 \}$$
  
* **$\sigma$-algebra ($\mathcal{F}$):** Since the sample space is finite, which includes every possible subset of outcomes.
$$\mathcal{F}=\mathcal{P}(\Omega)$$

* **Probability Measure ($P$):** A function that assigns the given probabilities to the events in $\Omega$:

  $$P(\{\omega_0\}) = 0.50, \quad P(\{\omega_1\}) = 0.30, \quad P(\{\omega_2\}) = 0.15, \quad P(\{\omega_3\}) = 0.05$$

### **2.**
A random variable $X$ is a function $X: \Omega \to \mathbb{R}$. we define $X$ as the mapping that assigns the count of failures to each outcome:
$$X(\omega_i) = i \quad \text{for } i \in \{0, 1, 2, 3\}$$

### **3.**
The PMF, $f(x) = P(X = x)$, defines the probability of each discrete value. For our variable $X$, the PMF is:

| $x$ | P(X = x) |
|---|---|
| 0 | 0.50 |
| 1 | 0.30 |
| 2 | 0.15 |
| 3 | 0.05 |

In [ ]:
# 4.
import numpy as np

# Define values and probabilities
x = np.array([0, 1, 2, 3])
p = np.array([0.50, 0.30, 0.15, 0.05])

# E[X] = Σ x * P(X=x)
expectation = np.sum(x * p)

print(f"4. The Expectation E[X] is: {expectation}")

# 5.
# Var(X) = Σ (x - E[X])^2 * P(X=x)
variance = np.sum((x - expectation)**2 * p)

print(f"5. The Variance Var(X) is: {variance:.4f}")

# 6.
# P(x>=1) is the sum of the prob for x=1,2,3
a=np.sum(p[x >= 1])
# P(x>=2) is the sum of the prob for x=2,3
b=np.sum(p[x >= 2])

print(f"6. The Probability P(X >= 1) is: {a:.2f}")
print(f"   The Probability P(X >= 2) is: {b:.2f}")

4. The Expectation E[X] is: 0.75
5. The Variance Var(X) is: 0.7875
6. The Probability P(X >= 1) is: 0.50
   The Probability P(X >= 2) is: 0.20


### **7.**

* **Expectation ($E[X] = 0.75$):** This is the **long-term average**. If we observe this district over many days, the average number of failures per day will be approximately 0.75. Practically, this tells the utility company how many repairs they should prepare for on a typical day.
* **Variance ($\text{Var}(X) = 0.7875$):** This measures the **volatility or spread**. It describes how much the actual number of failures varies from the average of 0.75. A higher variance would mean the company faces high uncertainty (some days with many failures, some with none), requiring more flexible emergency staffing.

# Q4)

### **1.**
To model the proportion of defective items $X$:
* **Sample Space ($\Omega$):** Since $X$ represents a proportion, the set of all possible outcomes is the unit interval:
  $$\Omega = [0, 1]$$

* **$\sigma$-algebra ($\mathcal{F}$):** We use the **Borel $\sigma$-algebra** on the interval $[0, 1]$. This collection includes all intervals and their countable unions/intersections within $[0, 1]$.

* **Probability Measure ($P$):** Defined for any event $A \in \mathcal{F}$ by the integral of the density function:
  $$P(A) = \int_A f(x) \, dx$$

### **2.**
$X$ is defined as the identity mapping on our sample space:
$$X: \Omega \to [0, 1] \quad \text{where} \quad X(\omega) = \omega$$
This means for any outcome $\omega$, the random variable $X$ simply takes that value.

### **3.**
The PDF of $X$ is given as:
$$f(x) = \begin{cases} \frac{1}{B(2, 18)}x(1 - x)^{17}, & 0 < x < 1 \\ 0, & \text{otherwise} \end{cases}$$
where $B(2, 18)$ is the Beta function, which serves as a normalizing constant.

### **4.**
A function is a valid PDF if it is non-negative and integrates to 1 over its support.

1.  **Non-negativity:** Since $x \in (0, 1)$, then $x > 0$ and $(1-x) > 0$. Therefore, $f(x) \ge 0$.

2.  **Total Integral:** By the definition of the Beta function $B(\alpha, \beta) = \int_0^1 t^{\alpha-1}(1-t)^{\beta-1} dt$:

  $$\int_0^1 f(x) \, dx = \int_0^1 \frac{x^{2-1}(1-x)^{18-1}}{B(2, 18)} \, dx = \frac{B(2, 18)}{B(2, 18)} = 1$$
Thus, $f$ is a valid PDF.

In [ ]:
import numpy as np
from scipy.stats import beta

alpha = 2
beta_param = 18

#5
# E[X] = alpha / (alpha + beta)
expectation = alpha / (alpha + beta_param)

#6
# Var(X) = (alpha * beta) / [(alpha + beta)^2 * (alpha + beta + 1)]
variance = (alpha * beta_param) / ((alpha + beta_param)**2 * (alpha + beta_param + 1))

print(f"5. Expectation E[X]: {expectation:.4f}")
print(f"6. Variance Var(X):  {variance:.6f} (approx {variance:.4f})")

from scipy.special import betainc

# 7. Probability at most 0.10 (Cumulative Distribution Function)
prob_at_most_10 = beta.cdf(0.10, alpha, beta_param)

# 8. Probability between 0.05 and 0.15
prob_between_05_15 = beta.cdf(0.15, alpha, beta_param) - beta.cdf(0.05, alpha, beta_param)

print(f"7. P(X <= 0.10):        {prob_at_most_10:.4f}")
print(f"8. P(0.05 <= X <= 0.15): {prob_between_05_15:.4f}")

5. Expectation E[X]: 0.1000
6. Variance Var(X):  0.004286 (approx 0.0043)
7. P(X <= 0.10):        0.5797
8. P(0.05 <= X <= 0.15): 0.5562


### **9.**
The Beta distribution is ideal for this situation because:
* **Bounded Support:** It is naturally defined on the interval $[0, 1]$, making it mathematically perfect for modeling **proportions, probabilities, or percentages**.
* **Flexibility:** By adjusting $\alpha$ and $\beta$, the distribution can represent various shapes, allowing it to model different manufacturing quality levels accurately.

### **10.**
* **Expectation ($E[X] = 0.10$):** This represents the **average defective proportion**. On an average production day, the company expects 10% of the items to be defective.

* **Variance ($\text{Var}(X) \approx 0.0043$):** This represents the **consistency of the process**.
    * A **low variance** indicates a stable process where the defect rate stays very close to 10% every day.
    * A **high variance** would suggest an unstable process where some days have very few defects and other days have many, making quality control difficult to predict.

# Q5)

### **1.**
The support of a random variable is the set of values for which the probability density function is non-zero. For this triangular distribution:
$$\text{Support} = [a, b] = [4, 10]$$
This means the installation will complete between 4 and 10 hours.

### **2.**
The general PDF $f_X(x)$ for a triangular distribution is piecewise. Given $a=4, b=10, c=6$:
- $b - a = 6$
- $c - a = 2$
- $b - c = 4$

The PDF is:
$$f_X(x) = \begin{cases} \frac{2(x-4)}{(10-4)(6-4)} = \frac{x-4}{6} & \text{for } 4 \le x \le 6 \\ \frac{2(10-x)}{(10-4)(10-6)} = \frac{10-x}{12} & \text{for } 6 < x \le 10 \\ 0 & \text{otherwise} \end{cases}$$

### **3.**
To verify, we calculate the area under the PDF curve. Since it is a triangle, using the geometric formula: $\text{Area} = \frac{1}{2} \times \text{base} \times \text{height}$.
* **Base:** $b - a = 10 - 4 = 6$
* **Height:** The value of the PDF at the mode ($c=6$):
  $$f_X(6) = \frac{6-4}{6} = \frac{2}{6} = \frac{1}{3}$$
* **Total Area:** $\frac{1}{2} \times 6 \times \frac{1}{3} = 1$

The integral over the support is 1, confirming it is a valid PDF.

In [ ]:
a, b, c = 4, 10, 6

#4.
# Expectation: E[X] = (a + b + c) / 3
expectation = (a + b + c) / 3

#5.
# Variance: Var(X) = (a^2 + b^2 + c^2 - ab - ac - bc) / 18
variance = (a**2 + b**2 + c**2 - a*b - a*c - b*c) / 18

print(f"4. Expectation E[X]: {expectation:.4f} hours")
print(f"5. Variance Var(X):  {variance:.4f}")

4. Expectation E[X]: 6.6667 hours
5. Variance Var(X):  1.5556


### **6.**
* **$a = 4$**: The **Optimistic** time. The absolute minimum time required if everything goes perfectly.
* **$b = 10$**: The **Pessimistic** time. The maximum time required, accounting for potential delays.
* **$c = 6$**: The **Most Likely** time. The value that would occur most frequently if the task were repeated many times.

### **7.**
It is highly effective when historical data is scarce but "expert judgment" is available. Unlike a Normal distribution, it has fixed boundaries (no infinite tails) and can represent **skewed** data (in this case, it's skewed to the right because the mode is closer to the minimum than the maximum).

In [ ]:
import numpy as np

def triangular_cdf(x, a, b, c):
    if x <= a: return 0
    if a < x <= c: return (x - a)**2 / ((b - a) * (c - a))
    if c < x < b: return 1 - (b - x)**2 / ((b - a) * (b - c))
    return 1

# 8. P(X <= 5)
prob_at_most_5 = triangular_cdf(5, a, b, c)

# 9. P(5 <= X <= 8)
prob_between_5_and_8 = triangular_cdf(8, a, b, c) - triangular_cdf(5, a, b, c)

print(f"8. P(X <= 5): {prob_at_most_5:.4f}")
print(f"9. P(5 <= X <= 8): {prob_between_5_and_8:.4f}")

8. P(X <= 5): 0.0833
9. P(5 <= X <= 8): 0.7500


### **10.**
* **Project Planning:** Managers can use the Expectation ($6.67$ hours) rather than just the mode ($6$ hours) to create more realistic schedules. The probabilities help determine "buffer" times. For example, knowing there is only an 8.3% chance ($P(X \le 5)$) of finishing in under 5 hours helps set expectations.
* **Sensitivity Analysis:** By adjusting $a, b,$ or $c$ (e.g., "What if the worst-case scenario $b$ increases to 15?"), planners can see how sensitive the total project timeline is to specific risks or delays in the generator installation phase.

# Q7)

In a PERT (Program Evaluation and Review Technique) distribution, the expected value (mean) $E[X]$ is calculated using the formula:
$$E[X] = \frac{a + 4c + b}{6}$$

Where:
* $a$ = Minimum time (optimistic)
* $b$ = Maximum time (pessimistic)
* $c$ = Most likely time (mode)

We are given:
* $a = 5$
* $c = 11$
* The constraint that $E[X] = c$ (The expected time matches the most likely time).

### **1.**
To find $b$, we set the expectation formula equal to $c$:
$$11 = \frac{5 + 4(11) + b}{6}$$

### **2.**

The variance of a PERT distribution is generally estimated by the formula:
$$\text{Var}(X) = \left( \frac{b - a}{6} \right)^2$$

**Observation on Symmetry:**
When the expected value $E[X]$ is equal to the most likely value $c$, the distribution becomes **perfectly symmetric**.
* Distance from min to mode: $c - a = 11 - 5 = 6$
* Distance from mode to max: $b - c = 17 - 11 = 6$

Since $c$ is exactly in the middle of $a$ and $b$, the "skewness" of the task disappears. In this specific setting, the variance is calculated directly from the full range of possible outcomes divided into 6 standard deviations.

In [ ]:
# Constants given in the problem
a = 5
c = 11

# Requirement: E[X] = c
# c = (a + 4*c + b) / 6
# Rearranging for b: b = 2*c - a

b = 2 * c - a

print(f"1. The value of the maximum time b is: {b} days")

# Calculating the variance
# Formula: Var(X) = ((b - a) / 6)**2

variance = ((b - a) / 6)**2

print(f"2. The variance Var(X) is: {variance}")
print(f"   The standard deviation (sigma) is: {variance**0.5}")

1. The value of the maximum time b is: 17 days
2. The variance Var(X) is: 4.0
   The standard deviation (sigma) is: 2.0
